dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/GeneratorPVAlphaSum.mo

```diff
@@ -0,0 +1,73 @@
+    parameter Types.ActivePowerPu PMinPu "Minimum active power in p.u (base SnRef)";
+    parameter Types.ActivePowerPu PMaxPu "Maximum active power in p.u (base SnRef)";
+    parameter Types.ReactivePowerPu QMinPu  "Minimum reactive power in p.u (base SnRef)";
+    parameter Types.ReactivePowerPu QMaxPu  "Maximum reactive power in p.u (base SnRef)";
+    parameter Types.ActivePowerPu PAuxHVPu = 0 "Active power consumption of the high voltage auxiliary used to recalculate the P and Q limits when the auxiliaries are not explicitly modeled (base SnRef)";
+    parameter Types.ActivePowerPu QAuxHVPu = 0 "Reactive power consumption of the high voltage auxiliary used to recalculate the P and Q limits when the auxiliaries are not explicitly modeled (base SnRef)";
+    parameter Types.ActivePowerPu PAuxLVPu = 0 "Active power consumption of the low voltage auxiliary used to recalculate the P and Q limits when the auxiliaries are not explicitly modeled  (base SnRef)";
+    parameter Types.ActivePowerPu QAuxLVPu = 0 "Reactive power consumption of the low voltage auxiliary used to recalculate the P and Q limits when the auxiliaries are not explicitly modeled  (base SnRef)";
+    final parameter Types.ActivePowerPu PMinAuxPu = PMinPu - PAuxHVPu - PAuxLVPu "Minimum active power in p.u when taking into account the auxiliaries (base SnRef)";
+    final parameter Types.ActivePowerPu PMaxAuxPu = PMaxPu - PAuxHVPu - PAuxLVPu "Maximum active power in p.u when taking into account the auxiliaries (base SnRef)";
+    final parameter Types.ReactivePowerPu QMinAuxPu = QMinPu - QAuxHVPu - QAuxLVPu "Minimum reactive power in p.u when taking into account the auxiliaries (base SnRef)";
+    final parameter Types.ReactivePowerPu QMaxAuxPu = QMaxPu - QAuxHVPu - QAuxLVPu "Maximum reactive power in p.u when taking into account the auxiliaries (base SnRef)";
+
+    parameter Real AlphaRaw "Reference value of the participation of the considered generator in the frequency regulation";
+
+    Connectors.ImPin N "Signal to change the active power reference setpoint of all the generators in the system in p.u (base SnRef)";
+    Connectors.ZPin Alpha "Participation of the considered generator in the frequency regulation";
+    Connectors.ZPin AlphaSum "Sum of all the participations of all generators in the frequency regulation";
+    Connectors.ZPin URefPu (value(start = U0Pu)) "Voltage regulation set point in p.u (base UNom)";
+
+  protected
+
+    Types.ActivePowerPu PGenRawPu (start = PGen0Pu) "Active power generation without taking limits into account in p.u (base SnRef) (generator convention)";
+
+  equation
+
+    if running.value then
+      PGenRawPu = PGen0Pu + (AlphaRaw * KGover / AlphaSum.value) * N.value;
+      PGenPu = if PGenRawPu >= PMaxAuxPu then PMaxAuxPu elseif PGenRawPu <= PMinAuxPu then PMinAuxPu else PGenRawPu;
+
+      if QGenPu >= QMaxAuxPu and (pre(QGenPu)<> QMaxAuxPu or UPu < URefPu.value) then
```

Comment:
I don't understand the condition in parenthesis: could you explain both terms?

Reply:
These conditions are used to make the generator go out of QMaxAuxPu or QMinAuxPu. If not used, the generator can be blocked at QMax or QMin and never go back even if the voltage goes higher (or lower) than its setpoint.

Reply:
If you use them as conditions to go back to the non-limited state, I would rather write them in the condition linked to the non-limited equation. More specifically, usually we use two loops: a when condition used to say if we are in standard mode or if we are limited and used to write in the timeline, and a if condition using the aforementionned state. Have a look in GeneratorPV/PQ and tell me what you think about this implementation.

Reply:
I think it would work but given that we don't want to use a timeline in the steady-state simulator, maybe it add variables and complexity that won't be used ? What do you think ?

Reply:
I'm not completely sure but without overthinking, at first glance, if you have an implementation like this (without timeline entries):

when A then
 statusA
elsewhen B then
 statusB
end when;

X = if statusA Y1 elseif statusbB Y2 else Y3;

It would even be more efficient than the current implementation because the condition in the when will only be tested during root and mode evaluations (which appears only once at each time-step) and the if will just be a pointer to the evaluation condition done into the when (not really doing the evaluation), whereas in your case, it could be done at each residual evaluation (which is between 1 and n by time steps).

But anyway, no big deal in performances I think here so adopt the clearest implementation.

Reply:
I would prefer to use the when loop as the conditions are all explicit and it is cleaner for the resolution also.

Reply:
Ok then if you both agree, I am doing this.